### DISCUSSION TITLE
Ask the dev

### DISCUSSION THREAD

**Discussion by m1kit** [m1kit/yalm-python#5] [General]
Hi @anshuldutt21, thank you for mentoring my GSoC project.

Currently I'm refactoring your code, and I have some points to ask you. Tell me about your thoughts when you have time!

**Comment by m1kit** [comment]
What is difference between [NormalizeText#returnfinalstring_for_template](https://github.com/spdx/spdx_python_licensematching/blob/master/normalize_license_text/normalize_class.py#L39) and [NormalizeTemplate#normalize_template](https://github.com/spdx/spdx_python_licensematching/blob/master/compare_template_text/normalize_template_text.py#L15)?

**Reply by anshuldutt21** [reply]
Hey @m1kit. So 1 task of the library is to compare the template which is a prespecified format to the license text. So basically I have normalised the license text first in the "returnfinalstring_for_template" function. Here as opposed to the [returnfinalstring](https://github.com/spdx/spdx_python_licensematching/blob/77608fd0c37bb21f2a49f60296a27d357c98542b/normalize_license_text/normalize_class.py#L23) method, I have omitted some functions as "bullets and numbering" and  "removing license title" because they are already covered by the template rules. 
The normalize template function basically normalizes the spdx template format and converts it into a regex format by implementing functions such as replacing text(replaceable and omitable) by regex. 
After this we compare these normalized forms using regex match.

**Comment by m1kit** [comment]
Why did you [replace whitespaces with backquote](https://github.com/spdx/spdx_python_licensematching/blob/master/normalize_license_text/normalize_class.py#L162)?

**Reply by anshuldutt21** [reply]
Backquotes are useful to match as the whitespaces were creating problems initially in terms of the length of it. Moreover, Backquotes was easy to handle as regexes than whitespaces. Debugging becomes easy as well since you can clearly see if there are any mismatches because of the number of backquotes.

**Reply by m1kit** [reply]
I thought that whitespace and backquote lengths were always normalized to 1, but are they not?

**Comment by m1kit** [comment]
You used `returnfinalstring_for_template` instead of of `returnfinalstring` [here](https://github.com/spdx/spdx_python_licensematching/blob/master/compare_template_text/compare_template_text.py#L26). I think it's a typo, but if not, tell me why.

**Reply by anshuldutt21** [reply]
As I mentioned in the first question you are getting a bit confused over returnfinalstring and returnfinalstring_for_template. The former is used to normalise license texts if we need to match against another license text. Here we are taking a large set of functions that remove most of the unnecessary parts (Like Bullets and numbering). The returnfinalstring_for_template is more precise and only normalizes specific parts and doesn't operate on functions like bullets and numbering since these cases are covered by the license templates for which we are using a regex match to compare.

**Reply by m1kit** [reply]
Oh, I understand. As you said I thought `returnfinalstring_for_template` is for normalization of templates, but it is not limited to template text.

**Comment by anshuldutt21** [comment]
They are always normalised to a single character. But its easier to check
if there are 2 backquote characters than 2 whitespace characters in case we
are debugging.

**Reply by m1kit** [reply]
I think it is strange that the template regex is filled with back-quotes for many people, but I get what you said. I'll think about it again when it's my turn to debug.

**Comment by anshuldutt21** [comment]
Okay take your time to think about this. The main purpose is just to have a
single backtick between every 2 words except for the regex. Some templates
are written in a way that the <<var>> is cluttered without any spaces
before or after. That's the reason for the incorporation of backquotes in
template regex as well.
